Tasks/Gulp/gulptask.ts

```diff
@@ -2,118 +2,162 @@
+		tl.mkdirP(cwd);
+		tl.cd(cwd);
 
-tl.debug('check path : ' + gulp);
-if (!tl.exist(gulp)) {
-	tl.debug('not found global installed gulp, try to find gulp locally.');
-	var gt = tl.createToolRunner(tl.which('node', true));
-	var gulpjs = tl.getInput('gulpjs', true);
-	gulpjs = path.resolve(cwd, gulpjs);
-	tl.debug('check path : ' + gulpjs);
-	if (!tl.exist(gulpjs)) {
-		tl.setResult(tl.TaskResult.Failed, tl.loc('GulpNotInstalled', gulpjs));
-	}
-	gt.pathArg(gulpjs);
-}
-else {
-	var gt = tl.createToolRunner(gulp);
-}
+		if (!tl.exist(gulp)) {
+			var gt = tl.tool(tl.which('node', true));
+			var gulpjs = tl.getInput('gulpjs', true);
+			gulpjs = path.resolve(cwd, gulpjs);
+			tl.debug('check path : ' + gulpjs);
+			if (!tl.exist(gulpjs)) {
+				tl.setResult(tl.TaskResult.Failed, tl.loc('GulpNotInstalled', gulpjs));
+				return;
+			}
+			gt.arg(gulpjs);
+		}
+		else {
+			var gt = tl.tool(gulp);
+		}
 
-if (isCodeCoverageEnabled) {
-	var npm = tl.createToolRunner(tl.which('npm', true));
-	npm.argString('install istanbul');
-	var testFramework = tl.getInput('testFramework', true);
-	var srcFiles = tl.getInput('srcFiles', false);
-	var testSrc = tl.getPathInput('testFiles', true, false);
-	var istanbul = tl.createToolRunner(tl.which('node', true));
-	istanbul.arg('./node_modules/istanbul/lib/cli.js');
-	istanbul.argString('cover --report cobertura --report html');
-	if (srcFiles) {
-		istanbul.argString('-i .' + path.sep + path.join(srcFiles));
-	}
-	if (testFramework.toLowerCase() == 'jasmine') {
-		istanbul.argString('./node_modules/jasmine/bin/jasmine.js JASMINE_CONFIG_PATH=node_modules/jasmine/lib/examples/jasmine.json');
-	} else {
-		istanbul.arg('./node_modules/mocha/bin/_mocha');
-	}
-	istanbul.arg(testSrc);
-	var summaryFile = path.join(cwd, 'coverage/cobertura-coverage.xml');
-	var reportDirectory = path.join(cwd, 'coverage/');
-}
+		if (isCodeCoverageEnabled) {
+			var npm = tl.tool(tl.which('npm', true));
+			npm.line('install istanbul');
+			var testFramework = tl.getInput('testFramework', true);
+			var srcFiles = tl.getInput('srcFiles', false);
+			var testSrc = tl.getPathInput('testFiles', true, false);
+			var istanbul = tl.tool(tl.which('node', true));
+			istanbul.arg('./node_modules/istanbul/lib/cli.js');
+			istanbul.line('cover --report cobertura --report html');
+			if (srcFiles) {
+				istanbul.line('-i .' + path.sep + path.join(srcFiles));
+			}
+			if (testFramework.toLowerCase() == 'jasmine') {
+				istanbul.line('./node_modules/jasmine/bin/jasmine.js JASMINE_CONFIG_PATH=node_modules/jasmine/lib/examples/jasmine.json');
+			} else {
+				istanbul.arg('./node_modules/mocha/bin/_mocha');
+			}
+			istanbul.arg(testSrc);
+			var summaryFile = path.join(cwd, 'coverage/cobertura-coverage.xml');
+			var reportDirectory = path.join(cwd, 'coverage/');
+		}
+
+		// optional - no targets will concat nothing
+		gt.arg(tl.getDelimitedInput('targets', ' ', false));
+		gt.arg('--gulpfile');
+		gt.arg("temp/gulpfile.js");
```

Comment:
Have a comment explaining that this is a temporary name which will be overridden.